### ISSUE TITLE
Generate SVGs with links

### ISSUE THREAD

**Issue by nyurik** [vega/vega#655]
SVG graphs really need links on Wikipedia. For a static graph, downloading entire Vega runtime plus all the required data just to click on a link and go to another page is not very good for such a heavy traffic site, so the "open" command does not help.  Thanks!

**Comment by jheer** [comment]
SVG supports links by wrapping elements within an `a` tag that includes the `xlink:href` property. Unfortunately there is not a straightforward "clean" way to add this to Vega's SVG output. Vega generally assumes a one-to-one mapping from a data element directly to an SVG element. Wrapping an element within an `a` tag may not be so simple -- I'm not yet sure how much refactoring this would involve (let alone the disconnect with canvas rendering). That said, if this feature is only needed for static output, a special mode of the SVGStringRenderer might be fairly straightforward to add. (Or, even a simple SVG post-processing step.) Would something along these lines suffice for your use case?

Implementation issues aside, there is also a spec design question here. Currently links are specified using the `open` function within an expression. For more general links, it might be better to have a dedicated Vega encoding property for a hyperlink (e.g., `"link"` or `"href"`). In normal usage, this might be realized via an `open` call under the hood, but could be treated specially when generating static SVG string output.

**Comment by nyurik** [comment]
This is purely for the static SVG generation, not for the interactive (client-side) usage.

**Comment by g3o2** [comment]
An "href" encoding property would be an awesome addition.

**Comment by jheer** [comment]
FYI, we now have a working prototype of this feature in the [jh/href branch of vega-scenegraph](https://github.com/vega/vega-scenegraph/tree/jh/href). Scenegraph items with an `href` property defined (settable via an `href` encoding property in Vega), will act as live hyperlinks.

For browser-based Canvas and SVG renderers, links are implemented by "spoofing" an `a` tag. A parent-less DOM anchor (`a`) tag is dynamically created and then dispatched an event with the same properties as a triggering click event. This preserves the desired link behavior with respect to modifier keys (e.g., Command-Click to open in new tab). Currently, having an `href` property does _not_ automatically adjust the cursor; instead for Vega the `cursor` encoding property must be used.

For the stand-alone SVGStringRenderer, the generated SVG instead has embedded `a` tags that wrap the linked SVG elements. If the output SVG is viewed in a browser, these links _do_ automatically change cursor, as is the browser default.

In all cases, `href` URLs must pass through sanitization using the standard Vega loader libraries. Sanitization is an asynchronous process, just like image loading. For "live" SVG and Canvas rendering, sanitization is performed on-demand in response to a clicked link. For the SVGStringRenderer, on the initial render sanitization will be requested for all href links, and no links will be included until sanitization completes. Thus, use the `renderAsync` method (which Vega 3's `View.toSVG` method already uses under the hood!) to ensure the output SVG is complete.

If I don't receive any additional (corrective) feedback, this feature should be folded into the next Vega beta release!

**Comment by nyurik** [comment]
Jeff, that's an awesome news!

Is there a way to customize the <a> building? For example, is there a way
to change attributes, add a class or a rel=nofollow attributes, perhaps
based on the target URL or the context of the graph?

Thanks!

**Comment by jheer** [comment]
Adding _static_ custom attributes that are universally applied should be more-or-less straightforward (e.g., via some runtime config parameters).

Introducing customized generation of attributes would be a bit more involved. Perhaps the most direct route would be a new sanitization API call that returns a hash of attributes instead of just a URL string. This would either be an entirely new method (adding bloat) or a re-design of the existing sanitization API (requiring re-factoring across multiple modules). So, I would like to be convinced that this is absolutely necessary before moving ahead.

**Comment by nyurik** [comment]
Jeff, a very basic use case - Wikipedia has two types of links - internal
(to other wiki projects), and external (anywhere else). In order to prevent
abuse (spam links or increasing site Google ranking by linking to it), we
decorate them differently. There might are some additional rules for RTL
languages and non-HTML (file download) links.

An external link:
*<a rel="nofollow" class="external text"
href="http://www.gutenberg.org/author/Albert_Einstein
<http://www.gutenberg.org/author/Albert_Einstein>">Works by Albert
Einstein</a>*

An internal link:
*<a href="/wiki/Albert_Einstein" title="Albert Einstein">Albert
Einstein</a>*

**Comment by jheer** [comment]
Thanks, that makes sense. However, your examples all use HTML `a` tags (not SVG). While SVG 2 revises the SVG spec for `a` tags to match HTML, SVG 1.1 does not support most of these attributes.

Unfortunately SVG 2 support is still spotty at best and it appears Firefox (for example) does not yet support the updated links: https://developer.mozilla.org/en-US/docs/Web/SVG/SVG_2_support_in_Mozilla

In short, while we can add support for generating attributes it's not clear when you would really be able to leverage it in production. I think this also points to a possible need for vega-scenegraph to be able to explicitly target either SVG 1.1 or SVG 2.0.

**Comment by nyurik** [comment]
Jeff, we do not yet allow any SVGs in production due to security concerns.
Editors have been relying extensively on image maps and complex and hard to
maintain tools like EasyTimeline
<https://www.mediawiki.org/wiki/Extension:EasyTimeline> (scroll down to see
usage examples). One of ongoing requests is to let users replace
easy-timeline horrid mixture of HTML and links with Vega-generated,
clickable static images. SVGs seem to be a good solution. Another would be
to auto-generate an image map.

WRT SVG support, I suspect that "rel=nofollow" would be a requirement,
because even if browsers ignore it, I suspect Google's crawler might (or
will) pay more attention to it.  The class is probably less important at
the moment since styling SVG elements might not be relevant to
graph-generated links.

**Comment by jheer** [comment]
OK. I've added `href` functionality, and it has been released in `v3.0.0-beta.7`! Also, the vega-loader `sanitize` method now returns an object whose properties should be valid `a` tag attributes. The sanitized URL should be available under the `href` property (the SVGStringRenderer automatically maps this to `xlink:href` for now, all other attributes pass through unchanged).

**Comment by nyurik** [comment]
Awesome news, thanks @jheer !!!

**Comment by williamfzc** [comment]
What a good feature and that is what i need, thanks!
But I did not find any usable sample snippets.
I have tried and followed the documents (and maybe some issues and PRs), and I still got no xlink in my svg.

![image](https://user-images.githubusercontent.com/13421694/143209598-3969ffb3-163e-4bfb-b779-b80b9780b34b.png)

My purpose is adding a `href` on the text in svg, just like clicking `a.md` and jumping to its url.

Here is a part of my spec:

```js
{
      type: "text",
      from: { data: "tree" },
      encode: {
          enter: {
              text: { field: "title" },
              fontSize: { value: 24 },
              baseline: { value: "middle" },

              // ???
              href: "www.google.com",
          },
          update: {
              x: { field: "x" },
              y: { field: "y" },
              dx: { signal: "datum.children ? -7 : 7" },
              align: {
                  signal: "datum.children ? 'right' : 'left'",
              },
              opacity: { value: 1 },
          },
         
          // ???
          "href": "www.google.com"
      },

      // ???
      href: "www.google.com",
  },
```

Where should I place my `href` prop?
Thanks

**Comment by domoritz** [comment]
Try wrapping it in an object with a value property.
